### PR TITLE
Garbage collector for allocations

### PR DESCRIPTION
**Pull request by diptanu**
This PR introduces Garbage Collection to the Nomad client.

### CODE REVIEW

File: client/gc.go [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+		if err := totalResource.Add(alloc.Resources); err != nil {
+			return err
+		}
+	}
+
+	// If the host has enough free space to accomodate the new allocations then
+	// we don't need to garbage collect terminated allocations
+	if hostStats := a.statsCollector.Stats(); hostStats != nil {
+		var availableForAllocations uint64
+		if hostStats.AllocDirStats.Available < uint64(a.reservedDiskMB*MB) {
```
Comment (dadgar) [inline client/gc.go]: I think you can just remove all the multipliers you are doing

File: client/stats/host.go [hunk 2]
```diff
@@ -52,32 +54,41 @@ type DiskStats struct {
 	InodesUsedPercent float64
 }
 
+type NodeStatsCollector interface {
```
Comment (dadgar) [inline client/stats/host.go]: Comment on interface + methods

File: client/stats/host.go [hunk 3]
```diff
@@ -52,32 +54,41 @@ type DiskStats struct {
...
 	statsCalculator map[string]*HostCpuStatsCalculator
 	logger          *log.Logger
+	hostStats       *HostStats
+	hostStatsLock   sync.RWMutex
+	allocDir        string
 }
 
 // NewHostStatsCollector returns a HostStatsCollector
-func NewHostStatsCollector(logger *log.Logger) *HostStatsCollector {
+func NewHostStatsCollector(logger *log.Logger, allocDir string) *HostStatsCollector {
```
Comment (dadgar) [inline client/stats/host.go]: Comment on why the alloc dir is passed in

File: client/stats/host.go [hunk 4]
```diff
@@ -122,32 +133,61 @@ func (h *HostStatsCollector) Collect() (*HostStats, error) {
...
-		if math.IsNaN(ds.InodesUsedPercent) {
-			ds.InodesUsedPercent = 0.0
-		}
-		diskStats = append(diskStats, &ds)
+		ds := h.toDiskStats(usage, &partition)
+		diskStats = append(diskStats, ds)
 	}
 	hs.DiskStats = diskStats
 
+	usage, err := disk.Usage(h.allocDir)
```
Comment (dadgar) [inline client/stats/host.go]: Comment here

File: command/agent/alloc_endpoint.go [hunk 5]
```diff
@@ -79,11 +79,26 @@ func (s *HTTPServer) ClientAllocRequest(resp http.ResponseWriter, req *http.Requ
...
+func (s *HTTPServer) ClientGCRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	if s.agent.client == nil {
+		return nil, clientNotRunning
+	}
+	err := s.agent.Client().CollectAllAllocs()
+	return nil, err
+}
+
+func (s *HTTPServer) allocGC(allocID string, resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	err := s.agent.Client().CollectAllocation(allocID)
```
Comment (dadgar) [inline command/agent/alloc_endpoint.go]: Guard against this being called on a server

File: command/agent/alloc_endpoint.go [hunk 6]
```diff
@@ -79,11 +79,26 @@ func (s *HTTPServer) ClientAllocRequest(resp http.ResponseWriter, req *http.Requ
...
 	case "snapshot":
 		return s.allocSnapshot(allocID, resp, req)
+	case "gc":
+		return s.allocGC(allocID, resp, req)
 	}
 
 	return nil, CodedError(404, resourceNotFoundErr)
 }
 
+func (s *HTTPServer) ClientGCRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
```
Comment (dadgar) [inline command/agent/alloc_endpoint.go]: Unit tests for both

File: client/gc_test.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+	if err := gc.MarkForCollection(ar2); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Make stats collector report 200MB free out of which 20MB is reserved
+	statsCollector.availableValues = []uint64{200 * MB}
+	statsCollector.usedPercents = []float64{0}
+	statsCollector.inodePercents = []float64{0}
+
+	alloc := mock.Alloc()
```
Comment (dadgar) [inline client/gc_test.go]: Can you set the disk un the tests so it is obvious in the test and the tests become resilient to changes to the mock alloc

File: client/gc_test.go [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+		t.Fatalf("err: %v", err)
+	}
+
+	statsCollector.availableValues = []uint64{1000}
+	statsCollector.usedPercents = []float64{20}
+	statsCollector.inodePercents = []float64{10}
+
+	if err := gc.keepUsageBelowThreshold(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
```
Comment (dadgar) [inline client/gc_test.go]: Do these two test anything? Don't they need to check if the allocs have been popped

File: client/gc.go [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+		// allocations
+		var allocDirStats *stats.DiskStats
+		if err := a.statsCollector.Collect(); err == nil {
+			if hostStats := a.statsCollector.Stats(); hostStats != nil {
+				allocDirStats = hostStats.AllocDirStats
+			}
+		}
+
+		if allocDirStats != nil {
+			if allocDirStats.Available >= uint64(totalResource.DiskMB*1024*1024) {
```
Comment (dadgar) [inline client/gc.go]: 1024*1024

File: client/gc.go [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+	return gc
+}
+
+func (a *AllocGarbageCollector) run() {
+	ticker := time.NewTicker(gcInterval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := a.keepUsageBelowThreshold(); err != nil {
+				a.logger.Printf("[ERR] client: error GCing allocation: %v", err)
```
Comment (diptanu) [inline client/gc.go]: The error could be coming in from the stats collector and not necessarily while destroying a particular allocation runner.

File: client/gc.go [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+
+func (a *AllocGarbageCollector) Stop() {
+	a.shutdownCh <- struct{}{}
+}
+
+// Collect garbage collects a single allocation on a node
+func (a *AllocGarbageCollector) Collect(allocID string) error {
+	gcAlloc, err := a.allocRunners.Remove(allocID)
+	if err != nil {
+		return fmt.Errorf("unable to collect allocation %q: %v", allocID, err)
```
Comment (diptanu) [inline client/gc.go]: @dadgar The error line will have something like this as the reason - `alloc 123 not present`

File: client/gc.go [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+		ar := gcAlloc.allocRunner
+		a.logger.Printf("[INFO] client: garbage collecting alloc runner for alloc %q", ar.Alloc().ID)
+		ar.Destroy()
+	}
+	return nil
+}
+
+// MakeRoomFor garbage collects enough number of allocations in the terminal
+// state to make room for new allocations
+func (a *AllocGarbageCollector) MakeRoomFor(allocations []*structs.Allocation) error {
```
Comment (diptanu) [inline client/gc.go]: @dadgar Once the complexity in the client is removed around calling `addAlloc` from multiple places we can call this API before we call the addAlloc and make room for more than one allocation before we start them.

**Pull request merged**